add Language::Node.npm_install_args helper method

chrmoritz opened the pull request:
**Update: This PR is now ready for review. The first initial design (described below) was discarded because it was to resource heavy. The discussion about all current issues with node module based formulas can be found from this https://github.com/Homebrew/brew/pull/37#issuecomment-208381710 and the discussion about the current approve starts with this https://github.com/Homebrew/brew/pull/37#issuecomment-208882330.**

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully ran `brew tests` with your changes locally?

### About handling node module formulas

Inspired by the discussion in legacy-homebrew#50167 I started to work on a prototype of a `Language::Node` for supporting formulas for node modules (with cmd line tools) were all dependencies are managed by homebrew resources and not by npm as it's done now. Especially because I've believed that most of the concerns mentioned by @DomT4 in [this comment](https://github.com/Homebrew/legacy-homebrew/issues/50167#issuecomment-198751313) are no longer an issue now. As it turned out that is the case for all points except the fact that the node community is taking the unix philosophy literally resulting in a high average count of dependencies / resources (which I have underestimated a little I have to admit). So please let's start the discussion if we want to have such a support, where we want to have it (core vs tap) and who we want to realize it!

### How to try it out

I've created a tap for it over at [`chrmoritz/node`](https://github.com/chrmoritz/homebrew-node) so that you can easier try it out (I think this is now the best way to go for testing changes affecting both `brew` and `homebrew-core`).
`brew tap chrmoritz/node`
It also comes with a node based [cmd line tool](https://github.com/chrmoritz/homebrew-node-formulator) for generating homebrew node formulas in the proposed style. It will install the module globally with npm@3 to a /tmp folder and parses the dep tree output from `npm ls --long --json` to generate the formula. You can install with `brew install chrmoritz/node/node-formulator --HEAD`. Then type `node-formulator -h` for usage instruction (btw: sry for the name). Currently it only fully supports npm registry hosted dependencies and has no support for / is not tested with optional and bundled dependencies yet.

It works really great for small tools like [generate-json-schema](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/generate-json-schema.rb) (currently in homebrew-core), [rimraf](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/rimraf.rb), [coffee-script](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/coffee-script.rb) and [coffeelint](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/coffeelint.rb) or the [node-formulator](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/node-formulator.rb) itself.
It still works fine for tools with a about average amount of dependencies like [less](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/less.rb) and [node-sass](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/node-sass.rb).
But is sadly currently hits homebrew limitation (e.g. no concurrent resource fetching) on very dependency heavy modules like [azure-cli](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli.rb) or [autocode](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/autocode.rb) (both currently in homebrew-core).

### How does it work

The proposed changes consist of 2 parts.
1. Adding a new `Language::Node` which include all the module installation logic and some other helper function (e.g. major node version helper for checking bottle compatibility of native addons).
2. Adding a new `Resource::Node` which extends the classic `resource` by two new fields: `parent` (used to map the dependency tree structure) and `install` (which is used to patch to install instruction for native addons to not use precompiled binaries, and yes we may should consider renaming it to avoid confusions).

*As you might have seen: I'm not a ruby developer. I'd appreciate it if someone could point me into the right direction to get rid of the helper getters in the `Resource::Node` or give any other feedback to improve to code style.*

##### How it installs the node modules

The actual installation is done by the `Language::Node.node_modules_install resources, libexec/"node_modules"` method. Just to say it at the beginning: The final structure of the `node_modules` folder is only partly npm compatible but fully node module spec compliant. I've taken the time to optimize it for our 'one time installation and never updating a dependency without reinstalling everything'-use case. It was slightly inspired by the flat dependency structure of npm@3, but I've gone further to be 100% flat (yes no nested modules for version conflicts) by using versioned dependencies and symlinks to their parents slightly inspired by [pnpm](https://github.com/rstacruz/pnpm) (yes we don't have multiple copies of the same nested dependency, we just have multiple symlinks from our versioned dependency at the root level to them).

If you look at the generated formulas over at [chrmoritz/homebrew-node/Formula](https://github.com/chrmoritz/homebrew-node/tree/master/Formula) you will see that there are to types of node `resources`. The first block of alphabetically sorted `resources` consists of modules simply named by their name, which should get installed at the root level the same way npm@3 would do it. The second block consist of the versioned modules (in npm@3 they would be nested modules) which are named like `name@version`, which we install in a versioned style at the root level to (similar to the way pnpm handles their modules). Don't worry they can't get accidentally get picked up because every not url safe character (including `@`) is a reserved character for module names in the npm ecosystem. Now that we have them there at one place (without unnecessary copies) we can create symlinks to their parents `node_modules` folder with only their names so they get picked up everywhere where they are needed. The information about their parents is read from the new `resource` metadata field which is filed by my formula generator tool with right information generated from the dependency tree. (Please note: In some very rare cases root dependency can also have parents, namely if a modules X dep in the dependency tree requires the version of module A from the root level, but somewhere in the dep tree branch between X and the root a other version of A is requested by another module and symlinked. Without the symlink from the root version of A this other version would be picked up by X.)

**Handling native addons:** This is still quite some work in progress in the current version. But the idea is that we have to patch most install scrips to not download prebuild binaries and build them always from source instead to be morw aligned with homebrews philosophy of building everything from source. This is done by adding an override field to the `resource` metadata describing how to compile the native addon, which should be used instead of the default one found in the `scripts.install` section of the `package.json`.
These fields are automatically generated by my tool from a patched version of the default install script in case of [node-pre-gyp](https://www.npmjs.com/package/node-pre-gyp) or [prebuild](https://www.npmjs.com/package/prebuild) usage, but they need to be edited by hand for native addons using a custom solution for providing binaries (but most native addons should use one of the 2 supported solutions or not provide any binaries at all).

### Advantages over the current system
*(All italic points could be backborted to the current system without homebrew manged resources)*

* installing doesn't rely on network access at install time (assuming you did a fetch earlier) as all dependencies are managed as npm resources
* *installing the same version of a formula with `--build-from-source` will always result in the same dependency versions to be installed. Currently the dependency are only locked to semver ranges and reinstalling the same version of a formula a week later will result in newer dependency release to be picked up (can bew fixed within the current system by applying a homebrew provided `npm_shrinkwrap.json`)*
* better handling for native addons inside the dependency chain:
  * I've a system in place for patching the install instructions to tell tools like [node-pre-gyp](https://www.npmjs.com/package/node-pre-gyp) or [prebuild](https://www.npmjs.com/package/prebuild) to always build them from source instead of downloading a prebuild binary (which is currently done for `fibers` in `azure-cli`).
  * *I'm also adding a `pour_bottle?` requirement for a specific node major version for formulas with a native addon in it's dependency chain, because the bottle won't be compatible with a node version which doesn't match the major version number with the node version used to build the bottle. (Don't forget the revision `azure-cli` with the upgrade to node v5.)*
  * the native addon compilation is independent of npm, so that it will work for users with a broken npm too (Why are still so many people playing russian roulett with their npm installation by running `npm -g update npm` instead of `npm -g install npm@latest`?). 
* *I'm reading the `package.json`s `bin` section for generating the `bin.install_symlink` code because the command name does not have to match the js file name (Just installing symlinks for everything in bin will not always do the correct thing. They don't even have to be located in the bin folder.)*
* a node module spec compliant 100% flat dependency structure with no duplication: I use versioned dependency (inspired by pnpm) instead of nested dependency to resolve version conflicts and symlink them to their parents `node_modules` folder instead of copying then multiple times over there. (We can do this more efficient than npm here because we don't need to support the 'updating only a specific dependency'-scenario, because every time we want to update a `resource` we have to reinstall the complete formula, which should happen only in a few exceptional cases outside of a formula version upgrade anyway.)
* Formula generation can be done fully automated with a simple command line tool which fills everything out except for the test. (You may have to edit some fields read from the `package.json` like `desc` by hand through if they doesn't match homebrews style.)

### Disadvantages of the proposed system
*these are caused by the high amount of dependencies / resources for which homebrew wasn't really designed for*

* long homebrew formulas (over 1000 lines for very dependency heavy modules)
* if building from source: long resources fetch time for dependency heavy modules (until homebrew implements concurrent resource fetching)

### Summary

As I've sad before: Don't see this as a finished PR. See it more like an prototype, which should start a discussion about how we can improve support for node module based formulas in homebrew.
I would appreciate any feedback about how the current approve can be evolved and how this can / should be supported by homebrew before I invest more time in it. 
I'm also prepared that you might say that the amount of resources is a show stopper for you (I wouldn't be disappointed if thats the case at this point, but I would like to get some feedback saying that this has a change before continuing working on it.). 
And sorry for writing such a long text ;-).

DomT4 commented:
I'll leave some thoughts here tomorrow, apologies for the silence so far.

chrmoritz commented:
No problem - I know that is a quite big topic to go through.

At this point the most important thing for me to know is, if this project will fail on the high number of `resources` it produces or if it has a change and it's worth to invest more time into it. Also some review of the Ruby part would help, because my Ruby skills are still quite limited. But if you want you can postpone the review of the node part of this (and over 50% of the text from above) until it is more finished.

Also note that native addon compilation is atm party broken, because npm relies on it's magic `.bin` folders for adding dep executable to the PATH at compilation time. But I've already an idea how to make the native addon compilation npm independent, which should solve this issue.
Should we go for a `depends_on :python => :build` build dependency if we detect a native addon or should we support system / user installed python (But in this case we can't guarantee that python in the PATH points to python 2.6/7 right?)?

mikemcquaid commented:
I'm personally in favour of reducing the number of `resource`s that need to be added manually or programmatically to formulae if there's an underlying package manager (i.e. `npm`) that doesn't just install everything from the tip of the `master` branch by default (I'm looking at you, `go` a few years ago), and allows packages to be vendored per-formula. I think `npm` meets both those criteria. The main area I personally see for improvement is pointing `npm` towards a cache that it can share between builds.

chrmoritz commented:
@mikemcquaid The design from this PR mostly addresses [this comment](https://github.com/Homebrew/legacy-homebrew/issues/50167#issuecomment-198751313) from @DomT4.
> * This means realistically if a user doesn't want the bottle, they need to be online during install and have sufficient connection strength to install npm-using formulae. This is an annoying problem for a package manager that does caching, and people have come to expect to be able to brew fetch something in advance and have it install offline later. It's a niche problem certainly, but an important niche.

But I agree that there are some improvements from this PR we could apply to our current node module formulas even if we don't want to increase the number of resources in favor of offline instability (after a fetch):
* doing global installs with a `--prefix` set to libexec instead of you current local installs
* better handling of native addons 
  * build dep on python
  * bottle pour requirement to same major node version
  * however avoiding pulling prebuilt binaries while continue using npm for installs is most likely not possible

Also npm already maintains a cache at `~/.npm` for modules, but we could separate it to use a custom independent cache for homebrew install using the `--cache=/path/to/cache` parameter.

chrmoritz commented:
@mikemcquaid Also note that with the current approve two subsequent `--build-from-source` installs of the same version of a node module based formula can produce different results, because the dependencies aren't usually locked to specific versions (only to semver ranges). IMHO this issue is comparable with the go head install issue.

It could be resolved by either shipping a homebrew provided `npm-shrinkwrap.json` for every node module based formula or going with this homebrew managed node module resources approve, which is sadly quite resource heavy.

mikemcquaid commented:
> - doing global installs with a --prefix set to libexec instead of you current local installs
> - Also npm already maintains a cache at ~/.npm for modules, but we could separate it to use a custom independent cache for homebrew install using the --cache=/path/to/cache parameter.

These both sound useful.

> Also note that with the current approve two subsequent --build-from-source installs of the same version of a node module based formulas can produce different results, because the dependencies aren't usually locked to specific versions (only to semver ranges). IMHO this issue is comparable with the go head install issue you've mentioned.

I think that's actually desirable; if there's an upstream security release I'd hope the buildsystem picks it up.

chrmoritz commented:
I've pushed an update to make native addon compilation independnet from npm to this PR an the example formula for azure-cli: https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli.rb#L1478-L1480

> I think that's actually desirable; if there's an upstream security release I'd hope the buildsystem picks it up.

But in this case we should do a revision which updates the bottle too. Currently only `--build-from-source` builds would benefit from the security update while the bottle would remain unpatched.

Also I think that you can except from over 95% of the current formulas in homebrew-core, that when you  install version x of formula y (from source) you will get exactly the same result as when you repeat it one week later (with still the same formula version).
But in the case of node module based formulas you will atm get a completely different dependency tree a week later (different dep versions and maybe also a different dep tree layout too because of resolving/adding new version conflicts through a dep update).

mikemcquaid commented:
> But in this case we should do a revision which updates the bottle too. Currently only --build-from-source builds would benefit from the security update while the bottle would remain unpatched.

I'm definitely open to bumping a revision when that happens.

> But in the case of node module based formulas you will atm get a completely different dependency tree a week later (different dep versions and maybe also a different dep tree layout too because of resolving/adding new version conflicts through a dep update).

Yeh, I'm sympathetic to that and open to solutions that don't involve extra work for either upstream or our formula contributors.

chrmoritz commented:
I think we can agree, that this PR on the one hand tries to solves a lot of issues with the current way we handle node modules based formulas, but unfortunately on the other hand introduces a new big issue through the big amount of dependencies it produces, which homebrew can't handle well enough atm. Also some :white_check_mark: (but not all :red_circle:) of the identified issues can be solved in another way too (without moving completely away from npm based installs). An example on how an improved formula can look like can be found [here for azure-cli](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli2.rb). I think we should move the discussion now more in the direction of these issues, which I've summarized bellow:

#### 1. Online requirement during installs (Refs [this comment](https://github.com/Homebrew/legacy-homebrew/issues/50167#issuecomment-198751313)) :red_circle:
* only fixable by homebrew managed resource dependencies as proposed by this PR

#### 2. Global vs local module installations :white_check_mark:
* Independent of what way we choose, we should switch our module installation from local style to global style (with setting a custom prefix to libexec) as in [this line]( https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli2.rb#L39)
* this brings our installations more in line with global installations through npm
* prunes the obsoletly installed devDependencies (which are unnecessarily installed with our current local installations) and
* uses npm for spec conform bin command name resolution (instead of relying on the developer to have a 1:1 mapping in a bin folder)

#### 3. Locking dependencies to specific versions :white_check_mark:
* atm when reinstalling the same version of the formula (from source) after a week you will get a completely different dependencies tree, because deps are only locked to semver ranges with the current approve
* this can either be fixed by homebrew managed resource deps (as proposed by this PR) or by providing a `npm-shrinkwrap.json` as [done here](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli2.rb#L32)
* this would require the generation of a new `npm-shrinkwrap.json` with every version upgrade by running `npm shrinkwrap` inside a global installation and updating the gist

#### 4. Handling of the npm cache :white_check_mark:
* atm the npm cache location gets overwritten by superenv to point to `buildpath/.brew_home/.npm` instead of `~/.npm`
* this results in over 10x higher build times than necessary making the current approve even less performant than the resource heavy approve proposed by this PR
* to fix this (while sticking with npm based installs) we need to tell npm the real location of it's global cache (and not our fake `buildpath/.brew_home/.npm` cache)
* a workaround for this can [be found here](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli2.rb#L35-L37) (Can this be improved somehow (can we access the old home?)?)

#### 5. NPM is not designed as a package manager for global installations :red_circle:
* NPM is optimized for the local installation use case and because of this is not creating optimized global installations
* the node module spec compliant format proposed in this PR is in 2 ways better than the global format produced by npm:
  * 100% flat (no nested dependencies, just symlinks to their locations)
  * no duplications (Npm produces multiple copies of the same version of a dependency if it's required at different nested places. In the proposed format we would just create multiple symlinks from our versioned dep to the places where it is needed, resulting in smaller installation sizes)

### 6. Handling of native addons (inside the dependency tree)

#### 6.1 Not pulling prebuild binaries :red_circle:
* Solutions like [node-pre-gyp](https://www.npmjs.com/package/node-pre-gyp) or [prebuild](https://www.npmjs.com/package/prebuild) for pulling prebuild binaries for native addons are not compatible with homebrews philosophy to build everything from source on the CI / in `--build-from-source` builds. That's why I have a system in place in this PR for patching the installation step of native addons to tell such tools to always build from source.
* Unfortunately it's is not possible to avoid pulling prebuild binaries of c-code for native addons if we stick with npm install based formula installations.

### 6.2 bottle compatibility with same major node version :white_check_mark:
* native addons compiled against a Node version are only compatible with other Node version with the same major version (other Node version are not ABI compatible)
* thats why we should consider adding a `pour_bottle?` requirement for native addons like [done here](https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli2.rb#L17-L22) for a specific node major version
* also a build time dep to python 2.6/2.7 is required too

mikemcquaid commented:
> Online requirement during installs

As described above I think this is unfortunate but a problem I'd like to be fixed (if we do) without requiring anything else of formula version creators/bumpers than "add URL, add checksum, done".

> Independent of what way we choose, we should switch our module installation from local style to global style (with setting a custom prefix to libexec) as in this line
> this brings our installations more in line with global installations through npm
> prunes the obsoletly installed devDependencies (which are unnecessarily installed with our current local installations) and

Sounds good :+1: 

> this can either be fixed by homebrew managed resource deps (as proposed by this PR) or by providing a npm-shrinkwrap.json as done here
> this would require the generation of a new npm-shrinkwrap.json with every version upgrade by running npm shrinkwrap inside a global installation and updating the gist

I think this is too much overhead for formula creators/updaters. I think if upstream don't provide files for this then it's fine for our versions to change each time you install.

> to fix this (while sticking with npm based installs) we need to tell npm the real location of it's global cache (and not our fake buildpath/.brew_home/.npm cache)
> a workaround for this can be found here (Can this be improved somehow (can we access the old home?)?)

Seems like a good idea :+1: However, the location should be somewhere under `HOMEBREW_CACHE`; we don't want to provide write-access to $HOME during install. We could consider copying files in there from `$HOME/.npm` at some point.

> the node module spec compliant format proposed in this PR is in 2 ways better than the global format produced by npm

I think this is something that's better taken upstream to `npm`.

> Unfortunately it's is not possible to avoid pulling prebuild binaries of c-code for native addons if we stick with npm install based formula installations.

Again, this is regrettable but not that serious, currently, and should be taken upstream to `npm`.


> native addons compiled against a Node version are only compatible with other Node version with the same major version (other Node version are not ABI compatible)
> thats why we should consider adding a pour_bottle? requirement for native addons like done here for a specific node major version

We don't actually need that requirement; `depends_on "node"` means that we can rely on the newest version of `node` always being installed before the package is installed (even if they are pinned). This would only be required if we had a `:node` or similar that allowed using Nodes from homebrew-versions.

---

Great write-up here, thanks for all your thought and work.

chrmoritz commented:
#### 4. Handling of the npm cache (+ 2. Global vs local module installations)

> Seems like a good idea :+1: However, the location should be somewhere under HOMEBREW_CACHE; we don't want to provide write-access to $HOME during install. We could consider copying files in there from $HOME/.npm at some point.

I think I might have an idea of what is causing the current `npm install` slowness. If no explicit cache path is set npm tries to resolve `~/.npm` both in it's main process and it's spawned child processes, but only it's main process is affected by the homebrew patched `ENV["HOME"]` by superenv. This results in a edge case because the cache is resolved to `buildpath/.brew_home/./npm` in the main process, but in the spawned child processes it's resolved to the actual `/Users/username/.npm`.

Setting the cache to somewhere inside `HOMEBREW_CACHE` doesn't seem to decrease the performance to much. These are the install times of `azure-cli` on a MBA 2012 (with a decent internet connection):
* with prefilled cache in `~/.npm`: ~45sec
* with empty cache in a tmp dir: ~50sec
* with unset cache causing edge case: aborted after 20 min (after running out of my 10gb free disk space)

Do you think the simple `HOMEBREW_CACHE/.npm` would be a good location or do you have something better in mind? Also do you think that this (and 2. global style module installation) would be worth creating a `Language::Node` helper similar to pythons `setup_install_args` which sets the cache (by creating a local `.npmrc` file with the cache entry set) and returning our npm install args for the global style (`"npm", "install", "--global", "--prefix=#{libexec}", "--verbose", buildpath`)?

#### 1. Online requirement during installs + 3. Locking dependencies to specific versions 

Ok, than I take this as a wont fix, because fixing it would add to much overhead.

#### 5. NPM is not designed as a package manager for global installations

> I think this is something that's better taken upstream to npm.

I disagree here. While it's not a common use case to update only a resource of a formula in homebrew, updating only specific dependencies is quite common in npm. What I meant here was, that we could optimize the module format more for the homebrew use case as it could be done for a npm use case. 
But I think it's like 1 and 3 something we can't improve without adding to much overhead.

#### 6.1 Native addons: Not pulling prebuild binaries

> Again, this is regrettable but not that serious, currently, and should be taken upstream to npm.

Npm official only supports building native addons from source. These are both 3rd party solutions and it might be worth asking them to add a environment variable to force source builds, which we could set.

#### 6.2 bottle compatibility with same major node version

> We don't actually need that requirement; depends_on "node" means that we can rely on the newest version of node always being installed before the package is installed (even if they are pinned). This would only be required if we had a :node or similar that allowed using Nodes from homebrew-versions.

Agreed. What we would really need to improve this is a way to enforce them to be executed with a specific node major version, which would require something like a major version specific opt prefix. So I guess this is also something which is not easily fixable.

mikemcquaid commented:
> Do you think the simple HOMEBREW_CACHE/.npm would be a good location or do you have something better in mind? Also do you think that this (and 2. global style module installation) would be worth creating a Language::Node helper similar to pythons setup_install_args which sets the cache (by creating a local .npmrc file with the cache entry set) and returning our npm install args for the global style ("npm", "install", "--global", "--prefix=#{libexec}", "--verbose", buildpath)?

Yeh, this all sounds pretty good. I'd probably go for `HOMEBREW_CACHE/npm` just to make the directory more obvious.

> Npm official only supports building native addons from source. These are both 3rd party solutions and it might be worth asking them to add an environment variable to force source builds, which we could set.

:+1: 

> Ok, than I take this as a wont fix, because fixing it would add to much overhead.
> But I think it's like 1 and 3 something we can't improve without adding to much overhead.
> Agreed. What we would really need to improve this is a way to enforce them to be executed with a specific node major version, which would require something like a major version specific opt prefix. So I guess this is also something which is not easily fixable.

Yeh, sadly on these. Thanks for understanding.

chrmoritz commented:
I've switched this PR to introduce the more simple `Language::Node.npm_install_args` helper method with this functionality:

* make sure that a working copy of npm and node-gyp from node's libexec is prepended to the PATH (to not rely of a user managed npm installations)
* sets the npm cache to HOMEBREW_CACHE/npm, which fixes issues under superenv resulting in long install times and high disk usage (https://github.com/Homebrew/brew/pull/37#issuecomment-208840366)
* returns npm install args for a global style module installation to libexec

A working copy of the `azure-cli` using this new helper method can be found here: https://github.com/chrmoritz/homebrew-node/blob/master/Formula/azure-cli.rb (`brew install chrmoritz/node/azure-cli`).

chrmoritz commented:
Updated formulas for all npm install based formulas in homebrew-core (except tegh which is a candidate for the boneyard imho) using this new helper method can be found here: https://github.com/Homebrew/homebrew-core/compare/master...chrmoritz:node_modules

mikemcquaid commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,19 @@
+      "cache=#{HOMEBREW_CACHE}/npm\n"
+    end
+
+    def self.npm_install_args(libexec)
```
Maybe call this `std_npm_install_args` and exclude `npm install`?

mikemcquaid commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,19 @@
+      # explicitly set npm's cache path to homebrew_npm_cache to fix issues
+      # under superenv (long build times, high disk space usage)
+      # https://github.com/Homebrew/brew/pull/37#issuecomment-208840366
+      Pathname.new(".brew_home/.npmrc").write npm_cache_config
```
May be worth checking if it already exists to avoid recreating unnecessarily.

mikemcquaid commented:
A couple of comments but I like the direction here :+1:

chrmoritz commented:
Addressed your comment and rebased on master.
I've also already changed the proposed formulas over at https://github.com/Homebrew/homebrew-core/compare/master...chrmoritz:node_modules to the new format, which has discarded you inline comments over there:
The current test of azure-cli is failing on node 6.0.0-rc.2, so native addon compatibility to node major version is already covered by it.
Also (node-)gyp requires Python 2.6 or 2.7. (Same as in the [v8 formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/v8.rb#L22))

chrmoritz commented:
Thinking this further I've added a second method named `Language::Node.local_npm_install_args`,  which can be used as a replacement for std_npm_install_args in formulas (like `elixirscript` and `kibana`), where npm install is only a part of the build step and global installations are therefore not suitable.

It still sets up the PATH for npm and our cache config but only returns the `—verbose` argument (as of now, but it might be extended in the future and therefore I've keeped it an array).

mikemcquaid commented:
@chrmoritz Looks good to me now but would like to get @DomT4 to 👍 before I merge.

DomT4 commented:
I'll try and look over and play with this tonight, but I'll get it done at the latest before Tuesday morning. One thing I can spot that'd be nice to get done is to get `brew cleanup` to clean the `npm`-cache in the same way we do the Java one.

DomT4 commented on Library/Homebrew/cleanup.rb:
```diff
@@ -50,7 +50,7 @@ def self.cleanup_cache
           next
         end
-        if path.basename.to_s == "java_cache" && path.directory?
+        if %w[java_cache npm_cache].include?(path.basename.to_s) && path.directory?
```
@UniqMartin may have feelings here. I think I got told off trying to do something similar to this with `include?` last time 😄. ("Told off" was tongue-in-cheek if that doesn't translate well across text!)

chrmoritz replied:
What would be the preferred way than? (A simple or would not fit on one line.)

DomT4 replied:
Possibly duplicating the existing `java_cache` block for `npm_cache` as well. Let's see what Martin has to say though, it's possible my memory is failing me, it happens alarmingly often.

UniqMartin replied:
I'm fine with the proposed version. Maybe now or when this array gets even longer, we might want to move that into a separate variable (or constant) with an intention-revealing name. But no strong feelings here.

DomT4 replied:
Alright, cool. Thanks for the review!

DomT4 commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,30 @@
+      # explicitly set npm's cache path to HOMEBREW_CACHE/npm_cache to fix
+      # issues under superenv (long build times, high disk space usage)
+      # https://github.com/Homebrew/brew/pull/37#issuecomment-208840366
+      npmrc = Pathname.new(".brew_home/.npmrc")
```
I'd be possibly tempted to use:

```ruby
npmrc = Pathname.new("#{ENV["HOME"]}/.npmrc")
```
Here, in case we ever decide to switch paths from `.brew_home` in future. It's unlikely, but not impossible.

DomT4 commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,30 @@
+      # managed ones in HOMEBREW_PREFIX/lib/node_modules which might be broken
+      ENV.prepend_path "PATH", "#{Formula["node"].opt_libexec}/npm/bin"
+      # explicitly set npm's cache path to HOMEBREW_CACHE/npm_cache to fix
+      # issues under superenv (long build times, high disk space usage)
```
This isn't actually a `superenv` thing. We override `HOME` in `formula` which means it happens regardless of env or sandbox status.

DomT4 commented:
### Summary:

This mostly looks reasonable, from testing. I like the way it ended up. Added a few comments but there's nothing enormous.

I agree with Mike in particular that avoiding adding a wad of resources to every `npm`-using formula was an important step to getting this 🚢-able; we've discovered that heavy resource use on things like Python formulae tends to result in people not updating them unless something explicitly breaks, which is a solid way to end up missing upstream resource security or bugfix releases.

It is an annoyance that many formulae can be fetched and then installed offline, and then some very much cannot, and that's something I'd like to eventually see if we can stamp out. I consider it a lesser evil  though to require someone to be online than implementing another system where resource security updates are potentially missed.

### Improvements:

One thing that would be nice to prevent is `.brew_home` being coped into every `npm`-using formula's `libexec/lib/node_modules/name` under the new system. It shouldn't cause any harm, but it could be confusing to users and is unnecessary clutter.

That directory will always contain at least a blob of folders followed by one file, the Python helper that makes sure Homebrew's `site-packages` for Python remain accessible to the system Python, as documented [here](https://github.com/Homebrew/brew/blob/acc9a7ca8554bc2413dee2d6d0f407b3a59c628c/Library/Homebrew/formula.rb#L1373-L1381).

### Things that'd be nice to handle one day:

One thing that does make me wonder if we've room for improvement are these comments in some formulae such as [here](https://github.com/chrmoritz/homebrew-core/blob/a568583597a58fa11e0b0a590861d886d205c39c/Formula/azure-cli.rb):
> needs a revision for every node major version bump (native addon)

I worry a bit that a lot of people are going to forget those lines exist and we'll end up doing revisions after-the-fact (if at all) rather than at-the-time.

It's probably overly complex and we may just have to live with this new mechanism being an improvement in stability over the status quo but it'd be nice if `audit` detected Node was being bumped more than x amount (which should be easy enough given Node's versioning system) and there weren't expected changes to formulae with that notation.

I don't think that idea is particularly workable immediately though, it's more of a "nice to have in future" than "let's get this done today".

chrmoritz commented:
I've addressed 2 of you comments and I'm now looking into removing `.brew_home` from `libexec`.

> I worry a bit that a lot of people are going to forget those lines exist and we'll end up doing revisions after-the-fact (if at all) rather than at-the-time.

As of now every node major version bump would trigger a test of all formulas which depend on node on the CI. If we would do a node major bump without adding a revision to `azure-cli` the test would fail because it would be compiled against node 5 but tested with node 6 for example. So this should be already covered by the current brew testbot implementation.

DomT4 commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,32 @@
+    def self.std_npm_install_args(libexec)
+      setup_npm_environment
+      # tell npm to not install .brew_home by adding it to .npmignore
+      open('.npmignore', 'a') { |f| f.write( "\n.brew_home\n") }
```
World's smallest nit, I suspect if we don't use double quotes around this element Rubocop will complain about it in text editors given our `StringLiterals` setting.

DomT4 replied:
To be clear: `'.npmignore', 'a'` - Those elements.

UniqMartin replied:
Are we certain that we're in the right directory when opening `.npmignore` with a relative path? Also, if we can be sure this file already exists, using [`Pathname#append_lines`](https://github.com/Homebrew/brew/blob/da34fba151ee33c1a2e14ab21ee0dc4ea451cc0f/Library/Homebrew/extend/pathname.rb#L145-L149) might be a neater way to do what is being done here, e.g. `Pathname.new(".npmignore").append_lines(".brew_home")`.

Just some thoughts, feel free to ignore if not applicable. (I've only skimmed the discussion in this long-ish PR.)

chrmoritz replied:
I will changed the quotes (silly copy and paste mistake).

@UniqMartin The npm command wouldn't work if we aren't in the right directory, so I think we can assume this here.
There is also no guarantee that a `.npmignore` file already exists. I'm relaying on the 'append if the file exists and create a new one if not'-functionality here.

UniqMartin replied:
@chrmoritz Thanks for the comment! Then ignore my suggestions and just fix the quotes. :smiley_cat:

DomT4 commented:
In "Dom is incredibly annoying" could you perhaps write a document [like this](https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/Python-for-Formula-Authors.md) _(without needing to be anywhere near as extensive)_ that serves as guidance for formulae authors moving to the new system?

Particularly on how and when to use each helper method, when to revision formulae that use native addons and a short example `Foo` formula.

That's my last point of concern here about moving the system across. If we can get that done I'm personally happy to ship.

UniqMartin commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,32 @@
+    def self.setup_npm_environment
+      # explicitly use our npm and node-gyp executables instead of the user
+      # managed ones in HOMEBREW_PREFIX/lib/node_modules which might be broken
+      ENV.prepend_path "PATH", "#{Formula["node"].opt_libexec}/npm/bin"
```
`opt_libexec` returns a `Pathname` instance. Thus, a slightly more readable version of this line would be:

```ruby
ENV.prepend_path "PATH", Formula["node"].opt_libexec/"npm/bin"
```

chrmoritz commented:
I've addressed all comments from above and squashed all fixup commits.
I will look into creating a documentation for the "annoying Don" tomorrow.

chrmoritz commented:
I've added a first draft of the [Node for formula authors](https://github.com/chrmoritz/brew/blob/nodelang/share/doc/homebrew/Node-for-Formula-Authors.md) documentation to this PR.

mikemcquaid commented:
Great work here, @chrmoritz. I'm happy to merge this if @DomT4 is.

xu-cheng commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,33 @@
+      open(".npmignore", "a") { |f| f.write( "\n.brew_home\n") }
+      # npm install args for global style module format installed into libexec
+      ["--verbose", "--global", "--prefix=#{libexec}", "."]
+    end
```
So if I call `std_npm_install_args` multiple times, I will end up with duplicated `Formula["node"].opt_libexec/"npm/bin"` in `PATH`, as well as `.npmignore` file duplicated entries.

Ultimately, it feels a little weird to have  `std_npm_install_args` and `local_npm_install_args` which supports to just return an argument list, to change any file system.

My suggest solutions are:
 * (a) create a separate method to setup environment. and call this method explicitly in formula.
 * (b) Or, make sure setup is only run once.

chrmoritz replied:
There should be no situation in which you have to call `std_npm_install_args` multiple times, especially not inside the same node module.

Through there might be a use case (currently not used in core) where calling `local_npm_install_args` multiple times might be needed: A formula which consists of multiple node modules and other stuff. In this case we would end up indeed with duplications in $PATH. I agree that we should guard the complete method and only the `.npmrc` creation against running multiple times.

Also the method names might not be the best describing names. Maybe something like `local_npm_install_config_and_args` would be better, through this would be already quite long for a method name.

mikemcquaid replied:
> I agree that we should guard the complete method and only the .npmrc creation against running multiple times.

This seems fine to me 👍

DomT4 commented on Library/Homebrew/language/node.rb:
```diff
@@ -0,0 +1,35 @@
+    def self.setup_npm_environment
+      npmrc = Pathname.new("#{ENV["HOME"]}/.npmrc")
+      # only run setup_npm_environment once per formula
+      return if npmrc.exist?
```
Just to check here, is there any chance a project ships an `.npmrc` in the tarball, which we then place in the `buildpath` and consequently this `return` is invoked before the `$PATH` is successfully modified below?

chrmoritz replied:
You mean how likely it is that a project ships a `.npmrc` file inside a `.brew_home` folder in their tarball? I would say quite unlikely and even than we would probably run into other issues with `.brew_home` too.

BTW: It's quite funny how a fix for a unlikely edge case of having to install more than one node module per formula has lead to a discussion about an even less likely edge case here.

DomT4 replied:
Oh, heh. This is what happens when I'm having a fairly long day; my brain ceases to function on anything above a "Keep Dom Alive" capacity. I briefly forgot we set it to `.brew_home` rather than just the `buildpath`. ` ¯\_(ツ)_/¯`

DomT4 commented on share/doc/homebrew/Node-for-Formula-Authors.md:
```diff
@@ -0,0 +1,115 @@
+
+# Download URL
+
+If the node module is also available on the npm registry, we prefer npm hosted release tarballs over github (or elsewhere) hosted source tarballs. The advantages of these tarballs are that they doesn't include the files from the `.npmignore` (such as tests) resulting in a smaller download size and that a possibly transpilation step is already done (e.g. no need to compile Coffee Script files as a ...
```
You're going to want to defenestrate me by the end of this review I'm sure, but it should be GitHub rather than `github`.

DomT4 replied:
`CoffeeScript` should probably be one word rather than two as well.

DomT4 commented on share/doc/homebrew/Node-for-Formula-Authors.md:
````diff
@@ -0,0 +1,115 @@
+depends_on "node"
+```
+
+If your formula requires to be executed with an older Node version you must vendor this older Node version as done in the [`kibana` formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/kibana.rb).
````
Can you peg the formulae references to a specific commit? That way if `kibana` or anything else stops using `node` in future the reference remains likely to be valid still.

DomT4 commented on share/doc/homebrew/Node-for-Formula-Authors.md:
```diff
@@ -0,0 +1,115 @@
+
+Node modules should be installed to `libexec`. This prevents the node modules from contaminating the global `node_modules`, which is important so that npm doesn't try to manage Homebrew-installed node modules.
+
+In the following we distinguish between 2 type of node module using formulas:
```
We prefer `formulae` over `formulas`.

DomT4 commented on share/doc/homebrew/Node-for-Formula-Authors.md:
```diff
@@ -0,0 +1,115 @@
+* formulas for standard node modules compatible with npm's global module format which should use [`std_npm_install_args`](#installing-global-style-modules-with-std_npm_install_args-to-libexec) (like [`azure-cli`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/azure-cli.rb) or [`autocode`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/autocode.rb)) and
+* formulas were the `npm install` step is only one of multiple not exclusively node related install steps (not compatible with npm's global module format) which have to use [`local_npm_install_args`](#installing-module-dependencies-locally-with-local_npm_install_args) (like [`elixirscript`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/elixirscript.rb) or [`kibana`](https://github ...
+
+Both methods have in common, that they are setting the correct environment for using npm inside Homebrew up and returning the arguments for invoking `npm install` for their specific use cases. This includes fixing an important edge case with the npm cache (caused by the way Homebrew patches the `$HOME` environment variable) by using our own custom `npm_cache` inside `HOMEBREW_CACHE`, which would  ...
```
It'd be nice to be a little more specific on our hostage-taking of `$HOME` so there's no implication it's a broad thing. Maybe something like:

```
(Caused by Homebrew's redirection of `$HOME` during the build and test process)
```
or such?

DomT4 commented on share/doc/homebrew/Node-for-Formula-Authors.md:
```diff
@@ -0,0 +1,115 @@
+    bin.install_symlink Dir["#{libexec}/bin/*"]
+  end
+
+  def test
```
`test do`. We haven't used `def test` for a fair chunk of time 😉.

DomT4 commented:
Mostly documentation notes at this point. Nearly there on getting this wrapped.

chrmoritz commented:
I've updated the documentation to address your comments and rebased on master.

DomT4 commented:
There's nothing I can see here that I've a problem with at this point. If nobody cries wolf over the next 12 hours or so I'll merge this then.

DomT4 commented:
Accidentally squashed it, which is a minor annoyance in this situation, but merged in https://github.com/Homebrew/brew/commit/bc98fd37882c64c896dc2243fcc6e129f170a32a.

Thanks for sticking with us on this review journey @chrmoritz, appreciate your patience and dedication in getting this done. If you wanted to follow-up by changing a few of the Node-using formulae in `homebrew/core` to use the new system that'd be ❤️ but entirely up to you.

chrmoritz commented:
Thx for merging this. I've submitted a rebase of the formulas (as proposed here) to core to make use of the new `Language::Node` in all node module based formulas (except `tegh`).

DomT4 commented:
Thanks! Spotted that and left a comment on the `tegh` situation.
